src/Confluent.SchemaRegistry/CachedSchemaRegistryClient.cs

```diff
@@ -0,0 +1,191 @@
+        private readonly Dictionary<string /*subject*/, Dictionary<string, int>> idBySchemaBySubject = new Dictionary<string, Dictionary<string, int>>();
+        private readonly Dictionary<string /*subject*/, Dictionary<int, string>> schemaByVersionBySubject = new Dictionary<string, Dictionary<int, string>>();
+        
+        /// <summary>
+        ///     The default timeout value for Schema Registry REST API calls.
+        /// </summary>
+        public const int DefaultTimeout = 10000;
+
+        /// <summary>
+        ///     The default maximum capacity of the local schema cache.
+        /// </summary>
+        public const int DefaultMaxCapacity = 1024;
+
+        /// <summary>
+        ///     Initialize a new instance of the SchemaRegistryClient class.
+        /// </summary>
+        /// <param name="config">
+        ///     Configuration properties.
+        /// </param>
+        public CachedSchemaRegistryClient(IEnumerable<KeyValuePair<string, object>> config)
+        {
+            if (config == null)
+            {
+                throw new ArgumentNullException("config properties must be specified.");
+            }
+
+            var schemaRegistryUrisMaybe = config.Where(prop => prop.Key.ToLower() == "schema.registry.url").FirstOrDefault();
+            if (schemaRegistryUrisMaybe.Value == null)
+            {
+                throw new ArgumentException("schema.registry.url configuration property must be specified.");
+            }
+            var schemaRegistryUris = (string)schemaRegistryUrisMaybe.Value;
+
+            var timeoutMsMaybe = config.Where(prop => prop.Key.ToLower() == "schema.registry.timeout.ms").FirstOrDefault();
+            var timeoutMs = timeoutMsMaybe.Value == null ? DefaultTimeout : (int)timeoutMsMaybe.Value;
+
+            var identityMapCapacityMaybe = config.Where(prop => prop.Key.ToLower() == "schema.registry.cache.capacity").FirstOrDefault();
+            this.identityMapCapacity = identityMapCapacityMaybe.Value == null ? DefaultMaxCapacity : (int)identityMapCapacityMaybe.Value;
+
+            this.restService = new RestService(schemaRegistryUris, timeoutMs);
+        }
+
+        /// <remarks>
+        ///     Ensure memory doesn't explode in the case of incorrect usage.
+        /// </remarks>
+        private void CheckIfCacheFull()
+        {
+            if (
+                this.schemaById.Count + 
+                this.schemaByVersionBySubject.Sum(x => x.Value.Count) +
+                this.idBySchemaBySubject.Sum(x => x.Value.Count)
+                    >= identityMapCapacity)
+            {
+                throw new OutOfMemoryException("Local schema cache maximum capacity exceeded");
+            }
+        }
+
+        /// <include file='include_docs.xml' path='API/Member[@name="ISchemaRegistryClient_GetIdAsync"]/*' />
+        public async Task<int> GetIdAsync(string subject, string schema)
```

Comment:
Using `string` here instead of `Avro.Schema` differs from the Java implementation and is a little bit controversial. It would be good to have another discussion about this with @mageshn or @ewencp.

Reply:
I prefer `Schema` for a few reasons:

1. If you already have a `Schema` object, you are forced to do the conversion, which seems a bit awkward.
2. Catch issues earlier. If the schema isn't even valid, we should know this before calling this method. I suspect this is expressed as a string here since you could potentially skip any validation and just pass along the string in the request to the server and assume it will validate. This could be fine since we do specifically have a special error code 42201 to indicate this issue, but I prefer catching the issue before ever making the request.
3. Related to 2, separation of concerns. I like keeping the methods here simpler if possible (i.e. no parsing in the method, we already know we have a valid schema so the types of possible errors are reduced).

That said, it kinda depends on what APIs the .net Avro library has. In the Java ones, people are frequently using SpecificRecord and the `Schema` is available directly as a `static` field, making it trivial to get at and natural to use. If the .net library was different and, e.g., just exposed it as a `string` instead, this might be a fairly natural fit.

Reply:
1. Yes. Actually we do have this in the `ISpecificRecord`'s in the dotnet case as well, and the serializer caches a `ToString`'d version of this so the conversion doesn't need to be done on every serialize. It's awkward but only very slightly and hidden.

2, 3. On the flip side, `Confluent.SchemaRegistry` currently has no dependency on `Apache.Avro` and accepting `Avro.Schema` would require this. You might argue most people using `Confluent.SchemaRegistry` will want this dependency anyway, but I also quite like that the current implementation is a more basic building block. I guess this is the main thing driving my opinion here (since the benefit to introducing the dependency although there is not huge). Still, the arguments the other way are perhaps stronger.

Two more things to note: 

1. This implementation is not susceptible to the reference comparison problems we saw in the Java version due to different instances of the Schema object representing the same schema being used. Not that that is inherently due to use of `Schema`, but in this implementation, there is no problem by default.

2. We can add an overload of these methods later if need be (numerous method overloads is common in .NET). I don't think i'll regret having the `string` overload, i.e. won't want to depreciate it.

Reply:
The other argument for using string is if we ever extend this to non-Avro as well. Although maybe string is worse in that case as you need some separate overrides for other languages.

re: the reference comparison thing -- that specifically has to do with caching. To properly resolve this, wouldn't an implementation here need to parse and then re-stringify the avro to ensure it has the canonicalized version? Otherwise schema strings with slightly different spacing would have effectively the same issue as, no?